### DISCUSSION TITLE
How to set complex properties on a target

### DISCUSSION THREAD

**Discussion by davidchisnall** [xmake-io/xmake#2768] [Q&A] [answered]
I would like to convey some structured data from a target to a rule.  I had hoped that I could write something like:

```
on_load(function (target)
    target:set("my property", { { key=value, key2=value2 }, { key=value3, key2=value4} })
end)
```

This doesn't raise an error, but it also doesn't appear to preserve the data.  I can set it as a JSON string, but it would be nice if I coudl use a Lua table literal so that I get parse errors early.

**Comment by waruqi** [comment, marked as answer]
please use `target:set_values` or set_values() and `target:values()`

**Reply by davidchisnall** [reply]
I don't think I'm using this correctly.  In my target, I have tried doing:

```
set_values("my property", { { key=value, key2=value2 }, { key=value3, key2=value4} })
```

And then in my rule, doing:

```
target:values("my property")
```

But I do not get the object back.  I can pass multiple arguments to `set_values` and end up with an array containing numbers or strings, but any object that I insert seems to be gone.  This is also true if I use `target:values_set` (`target:set_values` does not exist) in script scope.  

The information that I need to communicate is an array of objects, so mapping this to a set of strings and integers would be very error prone.  Using JSON works for now, but it feels very clunky in a language that has native support for objects to have to serialise and deserialise an array of objects as a string.

If I understand the code correctly, it appears that each `set` calls a different API based on a key, and `set_values` is in the `keyvalues` set.  There are some examples in comments such as `shflags` that appear to be successfully storing objects, so I'm not quite sure what is happening here.  Do I need to pass something to `set_values` / `values_set` to indicate that it should not try to coerce?  The code is quite light on comments, but I eventually discovered that adding `{expand=false}` as the last argument to `target:values_set` did the right thing.

This also works with `target:values_add` in the script scope but not with `add_values` in the description scope.

This almost works, but it has a couple of annoyances:

 - If I just add one object with `values_add`, I get a single object.  If I add another, I get an array of objects.  Lua arrays are just tables and so there's no simple way of differentiating between the two cases in the rule.  It would be nice if `add`, when given an object, would always give you a one-element array containing that object, though that's probably going to break some existing code.  Failing that, an accessor that can be used with `ipairs` to give an iterable list for a value would be convenient.
 - The objects always have a `__wrap_locked__` field in them.  If I create the array myself with `values_set`, then the array contains this, in addition to my objects, which can be confusing in iteration (though it appears that `ipairs` is happy to skip this one, so it's probably not important).

**Reply by waruqi** [reply]
It works for me.

```lua
add_rules("mode.debug", "mode.release")

target("testc")
    set_kind("binary")
    add_files("src/*.c")
    set_values("my property", { { key=1, key2=2 }, { key=2, key2=1} })
    on_load(function (target)
        print(target:values("my property"))
    end)
```

```console
$ xmake
{
  {
    key2 = 2,
    key = 1
  },
  {
    key2 = 1,
    key = 2
  }
}

[ 25%]: cache compiling.release src/main.c
[ 50%]: linking.release testc
[100%]: build ok!
```
